src/python/pants/core/util_rules/external_tool.py

```diff
@@ -175,6 +199,12 @@ def get_request(self, plat: Platform) -> ExternalToolRequest:
             f"{self.known_versions}"
         )
 
+    def check_version_constraint(self, version: str) -> None:
+        if self.constraints and not self.constraints.specifier.contains(version):  # type: ignore[attr-defined]
+            raise UnsupportedVersion(
+                f"Version {version} does not satisfy the version constraint {self.constraints}"
+            )
```

Comment:
right, this could be a logged warning, instead....

should we have this as an option, whether it is a hard requirement, or just a warning, and we could support it on a tool by tool basis, too..

Reply:
I think that whether it is a warning or an error, people need a way to override it. We should probably be thinking of this range as more like a warning sign that helps keep users from doing something accidentally, rather than like a barbed wire fence that prevents them from doing it.

Not sure how best to do that... maybe a new boolean option "--allow-potentially-incompatible" next to [`--version`](https://github.com/pantsbuild/pants/blob/bc46347cacea87618eeb3c65e1a7d0e5b01a0de7/src/python/pants/core/util_rules/external_tool.py#L87-L93)?

Reply:
I'll throw in a few ideas..
* `--force-version`
* `--ignore-version-constraints`
* `--i-wear-steel-hat-shoes-and-are-not-afraid-to-shoot-myself-in-the-foot`

Reply:
I think I like the last one most, as it gives the most flexibility, in allowing to add more alternatives in the future, if we end up with more use cases to fulfill..

Reply:
I do think erroring is the correct default behavior. Usually you run Pants upgrades through CI, and not everyone will notice warnings in CI.

> people need a way to override it. We should probably be thinking of this range as more like a warning sign that helps keep users from doing something accidentally, rather than like a barbed wire fence that prevents them from doing it.

Why? There are some things that _will_ break if you override it. I'm not convinced we need this until a user motivates a case. This is already a very advanced/niche feature, and I'm nervous about yet more options fatigue.

Reply:
Fine with erroring by default.

> Why? There are some things that _will_ break if you override it. I'm not convinced we need this until a user motivates a case.

Because we can't perfectly predict into the future which versions will break backwards compatibility: that makes it not a great idea to set a hard ceiling that someone can't change. On the "bottom of the range" side: we might bump the range because we know that one codepath (out of many) requires a particular version, but it's likely that other codepaths would continue to work.

> This is already a very advanced/niche feature, and I'm nervous about yet more options fatigue.

"Make easy things easy, and hard things possible."

Reply:
I agree with error by default, as it will be an easy config change to goto warning instead. But keeping the option to be able to continue regardless is valuable, as the incompatibilities may be in parts unused.

Also perhaps drop the allow/ignore alternative, as that would only remove the warning from the logs which may result in users forgetting about the "override" being potentially dangerous.

Reply:
I missed the semantic change from `ignore`, that when set to `allow`, it makes sense to log as info, so it is still visible that you are not adhering the constraints..